libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1361,26 +1394,154 @@ void IRGeneratorForStatements::appendAndOrOperatorCode(BinaryOperation const& _b
+	IRVariable value(_binOp);
+	define(value, _binOp.leftExpression());
 	if (op == Token::Or)
-		m_code << "if iszero(" << value << ") {\n";
+		m_code << "if iszero(" << value.name() << ") {\n";
 	else
-		m_code << "if " << value << " {\n";
+		m_code << "if " << value.name() << " {\n";
 	_binOp.rightExpression().accept(*this);
-	m_code << value << " := " + m_context.variable(_binOp.rightExpression()) << "\n";
+	assign(value, _binOp.rightExpression());
 	m_code << "}\n";
 }
 
-void IRGeneratorForStatements::setLValue(Expression const& _expression, unique_ptr<IRLValue> _lvalue)
+void IRGeneratorForStatements::assignCurrentLValue(IRVariable const& _value)
+{
+	solAssert(!!m_currentLValue, "No current LValue.");
+	auto const& type = m_currentLValue->type;
+	std::visit(
+		util::GenericVisitor{
+			[&](IRLValue::Storage const& _storage) {
+				std::optional<unsigned> offset;
+
+				if (std::holds_alternative<unsigned>(_storage.offset))
+					offset = std::get<unsigned>(_storage.offset);
+
+				IRVariable prepared(type, m_context.newYulVariable());
+				define(prepared, _value);
+
+				m_code <<
+					m_utils.updateStorageValueFunction(type, offset) <<
+					"(" <<
+					_storage.slot <<
+					(
+						std::holds_alternative<string>(_storage.offset) ?
+						(", " + std::get<string>(_storage.offset)) :
+						""
+					) <<
+					", " <<
+					prepared.commaSeparatedList() <<
+					")\n";
+			},
+			[&](IRLValue::Memory const& _memory) {
+				if (type.isValueType())
+				{
+					IRVariable prepared(type, m_context.newYulVariable());
+					define(prepared, _value);
+
+					// TODO: does this actually need cleanup?
+
+					if (_memory.byteArrayElement)
+					{
+						solAssert(type == *TypeProvider::byte(), "");
+						m_code << "mstore8(" + _memory.address + ", byte(0, " + prepared.commaSeparatedList() + "))\n";
+					}
+					else
+						m_code << m_utils.writeToMemoryFunction(type) <<
+							"(" <<
+							_memory.address <<
+							", " <<
+							prepared.commaSeparatedList() <<
+							")\n";
+				}
+				else
+				{
+					solAssert(type.sizeOnStack() == 1, "");
+					solAssert(dynamic_cast<ReferenceType const*>(&type), "");
+					m_code << "mstore(" + _memory.address + ", " + _value.name() + ")\n";
+				}
+			},
+			[&](IRLValue::Stack const& _stack) { assign(_stack.variable, _value); },
+			[&](IRLValue::Tuple const& _tuple) {
+				auto components = std::move(_tuple.components);
+				for (size_t i = 0; i < components.size(); i++)
+				{
+					size_t idx = components.size() - i - 1;
+					if (components[idx])
+					{
+						m_currentLValue.emplace(std::move(*components[idx]));
+						assignCurrentLValue(_value.tupleComponent(idx));
+					}
+				}
+			}
+		},
+		m_currentLValue->kind
+	);
+	m_currentLValue.reset();
+}
+
+IRVariable IRGeneratorForStatements::fetchCurrentLValue()
+{
+	solAssert(!!m_currentLValue, "No current LValue.");
+	auto const& type = m_currentLValue->type;
+	IRVariable result{type, m_context.newYulVariable()};
```

Comment:
So `newYulVariable` just reserves a prefix in the new model, right?

Reply:
Yep - and the actual variable names compute from this prefix plus the suffices from the type.